homeassistant/components/soma/cover.py

```diff
@@ -21,77 +38,179 @@ async def async_setup_entry(
     """Set up the Soma cover platform."""
 
     devices = hass.data[DOMAIN][DEVICES]
+    entities = []
+
+    for device in devices:
+        # Assume a shade device if the type is not present in the api response (Connect <2.2.6)
+        if ("type" in device.keys()) and (device["type"].lower() == "tilt"):
+            entities.append(SomaTilt(device, hass.data[DOMAIN][API]))
+        else:
+            entities.append(SomaShade(device, hass.data[DOMAIN][API]))
+
+    async_add_entities(entities, True)
+
+
+class SomaTilt(SomaEntity, CoverEntity):
+    """Representation of a Soma Tilt device."""
+
+    @property
+    def device_class(self):
+        """Return the class of this device."""
+        return DEVICE_CLASS_BLIND
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        supported_features = (
+            SUPPORT_OPEN_TILT
+            | SUPPORT_CLOSE_TILT
+            | SUPPORT_STOP_TILT
+            | SUPPORT_SET_TILT_POSITION
+        )
+        return supported_features
+
+    @property
+    def current_cover_tilt_position(self):
+        """Return the current cover tilt position."""
+        return self.current_position
+
+    @property
+    def is_closed(self):
+        """Return if the cover tilt is closed."""
+        return self.current_position == 0
+
+    def close_cover_tilt(self, **kwargs):
+        """Close the cover tilt."""
+        response = self.api.set_shade_position(self.device["mac"], 100)
+        if is_api_response_success(response):
+            self.current_position = 0
+        else:
+            _LOGGER.error(MSG_DEVICE_UNREACHABLE, self.device["name"], response["msg"])
+
+    def open_cover_tilt(self, **kwargs):
+        """Open the cover tilt."""
+        response = self.api.set_shade_position(self.device["mac"], -100)
+        if is_api_response_success(response):
+            self.current_position = 100
+        else:
+            _LOGGER.error(MSG_DEVICE_UNREACHABLE, self.device["name"], response["msg"])
+
+    def stop_cover_tilt(self, **kwargs):
+        """Stop the cover tilt."""
+        response = self.api.stop_shade(self.device["mac"])
+        if is_api_response_success(response):
+            # Set cover position to some value where up/down are both enabled
+            self.current_position = 50
```

Comment:
Should this call `self.schedule_update_ha_state()` to write the new state after its changed?

Reply:
For this I didn't change the way it was implemented before.
But I think you are right. The call to the api will stop the shade at a position X, and we should simply notify HA to update to the new state to get the current position instead of setting a value of 50. So I would use your suggestion instead of setting to the value 50. Make sense?

Reply:
It's fine to set the state optimistically like you have it but if you don't call self.schedule_update_ha_state() it will not be reflected in the state machine until the next poll and update

Reply:
Understood!
I have fixed this by creating a function "set_position" in the super class, handling the state update and call to "self.schedule_update_ha_state()".